src/java/main/org/apache/zookeeper/server/ZooKeeperServer.java

```diff
@@ -486,19 +475,57 @@ protected void startSessionTracker() {
         ((SessionTrackerImpl)sessionTracker).start();
     }
 
+    /**
+     * Sets the state of ZooKeeper server. After changing the state, this will
+     * notify the server state changes to registered shutdown handler, if any.
+     * <p>
+     * Following are the server state transitions:
+     * <li>During startup the server will be in INITIAL state.</li>
+     * <li>After the successful startup the server will set the state to
+     * RUNNING.</li>
+     * <li>Server will enter into ERROR state if it hits an internal error.
+     * {@link ZooKeeperServerListenerImpl} will notify critical resource error
+     * events.</li>
+     * <li>During shutdown the server will set the state to SHUTDOWN, which
+     * represents that the server is not running.</li>
+     *
+     * @param state new server state.
+     */
+    void setState(State state) {
+        this.state = state;
+        // Notify server state changes to the registered shutdown handler, if any.
+        if (zkShutdownHandler != null) {
```

Comment:
@fpj, The purpose of increasing the visibility is not for overriding the "enum State", that is to use "State" inside QuorumZooKeeperServer.java file. While overriding "#setState()" method it needs to pass "State" as an argument, like below:
    @Override
    protected void setState(State state) {
        this.state = state;
    }

Since "State" is defined inside ZooKeeperServer.java file with default visibility, it can't be used outside the package.

I think, we have agreed on all other points and this is the only pending part. I think uploading changes to this PR with proposed code changes will help us in discussion, shall I?